File: cracking/C/01.04.cracking.c
/*
Decide if two strings are anagrams
*/
#include <stdio.h>

int isAnagram(char *s, char *t) {
  if (s && t) {
    char sCount[256];
    int numUniqueCharsInS, numUniqueCharsTSoFar;
    printf("numUniqueCharsInS is: %d\n", numUniqueCharsInS);
    printf("numUniqueCharsTSoFar is: %d\n", numUniqueCharsTSoFar);
    while (*s) {
      if (!sCount[*s]) {
        numUniqueCharsInS++;
      }
      sCount[*s]++;
      s++;
    }
    printf("numUniqueCharsInS: %d\n", numUniqueCharsInS);
    while (*t) {
      if (!sCount[*t]) {
        return 0;
      }
      sCount[*t]--;
      if (!sCount[*t]) {
        ++numUniqueCharsTSoFar;
        printf("numUniqueCharsTSoFar: %d\n", numUniqueCharsTSoFar);
        if (numUniqueCharsInS == numUniqueCharsTSoFar) {
          return *(++t) == '\0';
        }
      }
      ++t;
    }
    return 1;
  } else {
    return 0;
  }
}

int main() {
  char s[10] = "anagram";
  char t[10] = "maaargn";
  printf("is anagram: %d\n", isAnagram(s, t));
  return 0;
}

File: cracking/Ruby/01.04.cracking.rb
=begin
Decide if two strings are anagrams or not
=end

def anagram(str1, str2)
  return false if str1.length != str2.length
  arr1 = Array.new(26, 0)
  arr2 = Array.new(26, 0)
  str1.downcase!; str2.downcase!
  str1.each_char do |char|
    arr1[char.ord - 'a'.ord] += 1
  end
  str2.each_char do |char|
    arr2[char.ord - 'a'.ord] += 1
  end
  (0..arr1.length - 1).each do |i|
    return false if arr1[i] != arr2[i]
  end
  true
end

a = "hello"; b = "elolh"
c = "ohlau"; d = "lhaue"
puts anagram(a, b)
puts anagram(c, d)
